components/request_otr/browser/request_otr_component_installer.cc

```diff
@@ -0,0 +1,73 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/request_otr/browser/request_otr_component_installer.h"
+
+#include <memory>
+#include <utility>
+
+#include "base/base_paths.h"
+#include "base/command_line.h"
+#include "base/containers/flat_set.h"
+#include "base/json/json_reader.h"
+#include "base/logging.h"
+#include "base/task/thread_pool.h"
+#include "base/types/expected.h"
+#include "brave/components/brave_component_updater/browser/dat_file_util.h"
+#include "brave/components/brave_component_updater/browser/local_data_files_service.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+
+using brave_component_updater::LocalDataFilesObserver;
+using brave_component_updater::LocalDataFilesService;
+
+namespace request_otr {
+
+const char kRequestOTRConfigFile[] = "request-otr.json";
+const char kRequestOTRConfigFileVersion[] = "1";
+
+RequestOTRComponentInstaller::RequestOTRComponentInstaller(
+    LocalDataFilesService* local_data_files_service)
+    : LocalDataFilesObserver(local_data_files_service) {}
+
+RequestOTRComponentInstaller::~RequestOTRComponentInstaller() = default;
+
+void RequestOTRComponentInstaller::LoadDirectlyFromResourcePath() {
+  base::FilePath dat_file_path =
+      resource_dir_.AppendASCII(kRequestOTRConfigFile);
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::MayBlock()},
+      base::BindOnce(&brave_component_updater::GetDATFileAsString,
+                     dat_file_path),
+      base::BindOnce(&RequestOTRComponentInstaller::OnDATFileDataReady,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void RequestOTRComponentInstaller::OnDATFileDataReady(
+    const std::string& contents) {
+  auto parsed_rules = RequestOTRRule::ParseRules(contents);
+  if (!parsed_rules.has_value()) {
+    LOG(WARNING) << parsed_rules.error();
+    return;
+  }
+  rules_.clear();
```

Comment:
is there a reason we need to keep a reference to these after install? It seems like wasted memory.

Reply:
I guess because of https://github.com/brave/brave-core/pull/17365/files#diff-80650dce890120bb2c6d20ffc7faa3ee04db3efb5a74df804736353d97e2c9b9R40 but I think it would be better to pass the rules in an observer method

Reply:
Other component-installer-based services also do this (e.g. debounce). The parsed rules and host cache have to live somewhere, in memory, forever. Where would you like them?

Reply:
Oh I see, once RequestOTRService can own them and the component installer can just pass in the unparsed JSON to the observer ready method. OK, done.